libsignal-service/src/gv2/api.rs

```diff
@@ -0,0 +1,156 @@
+    fn read(
+        &self,
+    ) -> Result<HashMap<i64, AuthCredentialResponse>, CredentialsCacheError>;
+    fn write(
+        &self,
+        value: &HashMap<i64, AuthCredentialResponse>,
+    ) -> Result<(), CredentialsCacheError>;
+}
+
+pub struct GroupsV2Api<S: PushService, C: CredentialsCache> {
+    push_service: S,
+    credentials_cache: C,
+    server_public_params: ServerPublicParams,
+}
+
+impl<S: PushService, C: CredentialsCache> GroupsV2Api<S, C> {
+    pub fn new(
+        push_service: S,
+        credentials_cache: C,
+        server_public_params: ServerPublicParams,
+    ) -> Self {
+        Self {
+            push_service,
+            credentials_cache,
+            server_public_params,
+        }
+    }
+
+    pub async fn get_authorization_for_today(
+        &mut self,
+        uuid: Uuid,
+        group_secret_params: GroupSecretParams,
+    ) -> Result<(String, String), ServiceError> {
+        let today = Self::current_time_days();
+        let mut cached_credentials = self.credentials_cache.read()?;
+        let auth_credential_response = if let Some(auth_credential_response) =
+            cached_credentials.remove(&today)
+        {
+            auth_credential_response
+        } else {
+            let mut new_credentials =
+                self.get_authorization(today).await?.parse();
+            self.credentials_cache.write(&new_credentials)?;
+            new_credentials.remove(&today).ok_or_else(|| {
+                ServiceError::ResponseError {
+                    reason:
+                        "credentials received did not contain requested day"
+                            .into(),
+                }
+            })?
+        };
+
+        self.get_authorization_string(
+            uuid,
+            group_secret_params,
+            auth_credential_response,
+            today as u32,
+        )
+    }
+
+    async fn get_authorization(
+        &mut self,
+        today: i64,
+    ) -> Result<CredentialResponse, ServiceError> {
+        let today_plus_7_days = today + 7;
+
+        let path =
+            format!("/v1/certificate/group/{}/{}", today, today_plus_7_days);
+
+        self.push_service
+            .get_json(Endpoint::Service, &path, None)
+            .await
+    }
+
+    fn current_time_days() -> i64 {
+        let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
+        let today = chrono::Duration::from_std(now).unwrap();
+        today.num_days()
+    }
+
+    fn get_authorization_string(
+        &self,
+        uuid: Uuid,
+        group_secret_params: GroupSecretParams,
+        credential_response: AuthCredentialResponse,
+        today: u32,
+    ) -> Result<(String, String), ServiceError> {
+        let auth_credential = self
+            .server_public_params
+            .receive_auth_credential(
+                *uuid.as_bytes(),
+                today,
+                &credential_response,
+            )
+            .unwrap();
```

Comment:
This sounds fallible, no? Otherwise, maybe document why this unwrap is safe, possibly by using .expect instead.